core/lib/src/local/form.rs

```diff
@@ -0,0 +1,222 @@
+    where
+        N: Into<NameBuf<'v>>,
+        V: AsRef<str>,
+    {
+        self.0
+            .push(LocalField::Value(name.into(), value.as_ref().to_owned()));
+        self
+    }
+
+    /// Adds all the `fields` to `self`.
+    pub fn fields<I, F>(mut self, fields: I) -> Self
+    where
+        I: Iterator<Item = F>,
+        F: Into<ValueField<'v>>,
+    {
+        fields.for_each(|field| {
+            let field = field.into();
+            self.0.push(LocalField::Value(
+                NameBuf::from(field.name.as_name()),
+                field.value.to_string(),
+            ));
+        });
+        self
+    }
+
+    /// A percent-encoded `name` and `value`.
+    pub fn raw_field(mut self, name: &'v RawStr, value: &'v RawStr) -> Self {
+        let decoded_name = name.percent_decode_lossy().into_owned();
+        let value = value.percent_decode_lossy().into_owned();
+        self.0
+            .push(LocalField::Value(NameBuf::from(decoded_name), value));
+        self
+    }
+
+    /// Adds a field for the file with the name `file_name`, content-type `ct`, and
+    /// file contents `data`.
+    pub fn file<N, V>(mut self, file_name: N, ct: ContentType, data: V) -> Self
+    where
+        N: Into<Option<&'v str>>,
+        V: AsRef<[u8]>,
+    {
+        self.0.push(LocalField::Data(
+            NameBuf::from("file"),
+            file_name.into(),
+            ct,
+            data.as_ref().into(),
+        ));
+        self
+    }
+
+    /// Add a data field with a content-type `ct` and binary `data`.
+    pub fn data<V>(mut self, ct: ContentType, data: V) -> Self
+    where
+        V: AsRef<[u8]>,
+    {
+        self.0.push(LocalField::Data(
+            NameBuf::from("file"),
+            None,
+            ct,
+            data.as_ref().into(),
+        ));
+        self
+    }
+
+    /// The full content-type for this form.
+    pub fn content_type(&self) -> ContentType {
+        if self.contains_data_field() {
+            return "multipart/form-data; boundary=X-BOUNDARY"
+                .parse::<ContentType>()
+                .unwrap();
+        }
+
+        ContentType::Form
+    }
+
+    /// The full body data for this form.
+    pub fn body_data(&self) -> Vec<u8> {
+        if self.contains_data_field() {
+            self.format_multipart()
+        } else {
+            self.format_simple()
+        }
+    }
+
+    fn contains_data_field(&self) -> bool {
+        self
+            .0
+            .iter()
+            .any(|field| matches!(field, LocalField::Data(..)))
+    }
+
+    fn format_simple(&self) -> Vec<u8> {
+        self.0.iter().fold(Vec::new(), |mut acc, field| {
+            match field {
+                LocalField::Value(name, value) => {
+                    acc.push(format!("{}={}", Uri::percent_encode(&format!("{}", name)), Uri::percent_encode(value)));
+                    return acc
+                },
+                _ => acc,
+            }
+        }).join("&").as_bytes().to_vec()
```

Comment:
Ideally we allocate a total of 1 `Vec<u8>` here, pass a pointer it around, and use `write!` to fill it up.

Here's an example of how to do that:

```rust
use std::io;

fn f(buf: &mut dyn io::Write) -> io::Result<()> {
    write!(buf, "hello, {}", "there")?;
    Ok(())
}

pub fn main() {
    let mut buf = vec![];
    let _ = f(&mut buf);
}
```

I'd like the this code to have _zero_ calls to `format!`, or `to_string()`, or similar.